### PR TITLE
chore (client): improved interface for the generic adapter

### PR DESCRIPTION
**Pull request by kevin-dp**
This PR improves the generic database adapter to require `_query` and `_run` methods instead of an `exec` method.

### CODE REVIEW

File: clients/typescript/src/drivers/generic/adapter.ts [hunk 1]
```diff
@@ -28,14 +27,19 @@ abstract class DatabaseAdapter
...
 
   /**
-   * @returns The number of rows modified by the last insert/update/delete query.
+   * Runs a single SQL query against the DB.
+   * The query should be read-only.
+   * @param stmt The SQL statement to execute
+   * @returns The rows read by the query.
    */
-  abstract getRowsModified(): number
+  abstract _query(stmt: Statement): Promise<Row[]>
```
Comment (icehaunter) [inline clients/typescript/src/drivers/generic/adapter.ts:42]: Does this assume no `RETURNING` clauses ever?

**Review by samwillis — APPROVED** [review]
@kevin-dp Looks good, I'm all for explicit readonly and write paths.

One thought, we merged #805 last week that introduced `rawQuery` and `unsafeExec` as the front end *enforced* read only and write apis. It may be a good idea to stick to the same naming, so `_exec` rather than `_run`.

File: clients/typescript/src/drivers/generic/adapter.ts (hunk 1 above)
Comment (kevin-dp) [reply clients/typescript/src/drivers/generic/adapter.ts:42]: If you're using a `RETURNING` clause you have to use `_query`. My understanding is that `query` is not only for read-only queries but is also for modification statements that return rows use the `RETURNING` clause. That's also what we do internally in the DAL (e.g. the DAL's `create` method does an `INSERT INTO` with a `RETURNING` clause and executes it using `query`).

**Comment by kevin-dp** [comment]
> One thought, we merged #805 last week that introduced `rawQuery` and `unsafeExec` as the front end _enforced_ read only and write apis. It may be a good idea to stick to the same naming, so `_exec` rather than `_run`.

This abstract class `DatabaseAdapter` is extended by developers to add support for a new SQLite driver.
These drivers typically provide `run` and `query` methods so i think that is a bit more appropriate here.
I think it's more important to stick to the naming of the drivers here (rather than DAL naming) because this is concerned with adding driver support (not related to the DAL).

File: clients/typescript/src/drivers/generic/adapter.ts (hunk 1 above)
Comment (icehaunter) [reply clients/typescript/src/drivers/generic/adapter.ts:42]: Yeah, that's the expected behaviour. If that's the case, remove `read-only` from the doc here pls :pray:

**Review by icehaunter — APPROVED** [review]

**Pull request merged**